### DISCUSSION TITLE
Disable function codes

### DISCUSSION THREAD

**Discussion by dlmoffett** [pymodbus-dev/pymodbus#1971] [Help] [answered]
Hey, I've got a funky application where, as a Server, I need to disable many standard function codes, and more importantly, return the `IllegalFunction` response if a client transmits a request for a disabled function code.

Specifically we are only supporting function codes `3` and `16`.

I've searched and cannot find any obvious/easy way of doing this.

**Comment by zufar-ik** [comment]
❤️

**Comment by janiversen** [comment]
The simulator datastore can do that. Alternatively you can patch the framer.

**Reply by dlmoffett** [reply]
I was going down the patch route but it wasn't immediately obvious to me what the framer should do in order to have the device actually send the illegal function response. 

- Is there an exception the framer should raise (I was having a hard time determining where any such exception gets caught and how it gets handled)?
- Or should it mangle request - replacing the function code with an obviously invalid one?
- Some other, more intelligent option?

**Reply by janiversen** [reply, marked as answer]
Well look at the code, that's about the best advice I can give you...the framer can reject a frame, manipulate a frame and/or generate responses, but of course you need to dig into the code to understand how it works.
